dg/tests.md

```diff
@@ -0,0 +1,166 @@
+
+### Spy
+
+`Spy` objects are a very important aspect of Javascript tests. They enable users to stub functions and keep track of useful information such as the number of calls to the function, parameters it has been called with, etc.
+
+For example, if function B is called within function A which is under test, the user is able to find out detailed information regarding how function B is called within A by creating a spy object of B. This allows testers to verify that B has indeed been called with the correct arguments correct number of times.
+
+Example of a `Spy` object in CATcher
+```
+describe('DescriptionComponent', () => {
+    beforeEach(() => {
+        ...
+    dialog = jasmine.createSpyObj('MatDialog', ['open']);
+    errorHandlingService = jasmine.createSpyObj('ErrorHandlingService', ['handleError']);
+    issueService = jasmine.createSpyObj('IssueService', ['getLatestIssue', 'updateIssue']);
+        ...
+    });
+
+    it('should update the form value correctly and emit an event when entering edit mode', () => {
+    const descriptionComponentEditState = spyOn(descriptionComponent.changeEditState, 'emit');
+
+    descriptionComponent.ngOnInit();
+    descriptionComponent.changeToEditMode();
+    expect(descriptionComponentEditState).toHaveBeenCalledTimes(1);
+    expect(descriptionComponent.issueDescriptionForm.value).toEqual({ description: thisIssue.description });
+    });
+    ...
+}
+
+```
+The test checks that the method `descriptionComponentEditState` has indeed been called once, and that the value of `issueDescriptionForm` within the `descriptionComponent` is the expected value after update.
+
+You are recommended to read the official Jasmine documentation (Links below) for more details and examples on spies and their functionalities.
+
+### Resources for Jasmine
+
+1. [Official Jasmine documentation](https://jasmine.github.io/api/3.6/global) : This is the official Jasmine documentation for Jasmine 3.6
+2. [Introduction to Jasmine 2.0](https://jasmine.github.io/2.0/introduction.html) : This is a good summary / introduction of Jasmine test features
+
+## Angular TestBed Utility
+
+Because the above Jasmine framework does not test the DOM, we require the Angular TestBed Utility functions to set up component tests for testing HTML / view changes of components in CATcher.
+
+Steps to set up component tests:
+1. Configure the testing module through `TestBed` function `configureTestingModule` with the corresponding component's settings
+2. Use `TestBed` function to create the component (fixture) to be tested
+3. Observe HTML changes in the fixture during testing of functions by querying HTML elements of the fixture
+
+Code snippet of a component test in CATcher
+```
+describe('AssigneeComponent', () => {
```

Comment:
I think we should minimise usage of code snippets that can change over time.
Moreover, this snippet doesn't show the actual tests (just the component setup).
How about we just link to the test file itself?

Reply:
That sounds better!